File: crates/api_clients/artcraft/artcraft_api_defs/src/generate/cost_estimate/estimate_image_cost.rs
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

use enums::common::generation::common_aspect_ratio::CommonAspectRatio;
use enums::common::generation::common_image_model::CommonImageModel;
use enums::common::generation::common_quality::CommonQuality;
use enums::common::generation::common_resolution::CommonResolution;
use enums::common::generation_provider::GenerationProvider;

pub const ESTIMATE_IMAGE_COST_PATH: &str = "/v1/generate/cost_estimate/image";

/// Request body for the image cost estimate endpoint.
#[derive(Serialize, Deserialize, ToSchema, Debug)]
pub struct EstimateImageCostRequest {
  /// The image model to estimate costs for.
  pub model: CommonImageModel,

  /// The provider to route the generation through.
  pub provider: GenerationProvider,

  /// The type of generation (determines whether input images are involved).
  /// This is a tagged enum, so it looks like:
  ///   "generation_mode": {"type": "text_to_image"}
  ///   "generation_mode": {"type": "image_edit", "count": 2}
  pub generation_mode: GenerationMode,

  /// Optional aspect ratio.
  pub aspect_ratio: Option<CommonAspectRatio>,

  /// Optional resolution.
  pub resolution: Option<CommonResolution>,

  // Optional quality.
  pub quality: Option<CommonQuality>,

  /// Number of images to generate in parallel.
  pub image_batch_count: Option<u16>,
}

/// Describes the type of image generation being requested.
#[derive(Serialize, Deserialize, ToSchema, Debug)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum GenerationMode {
  /// No input images — generate from prompt only.
  TextToImage,
  /// One or more input images are provided for editing/compositing.
  ImageEdit { count: u32 },
}

/// Response body for the image cost estimate endpoint.
#[derive(Serialize, Deserialize, ToSchema)]
pub struct EstimateImageCostResponse {
  pub success: bool,

  /// Estimated cost in credits.
  pub cost_in_credits: Option<u64>,

  /// Estimated cost in USD cents.
  pub cost_in_usd_cents: Option<u64>,

  /// Whether the generation is free for this user/plan.
  pub is_free: bool,

  /// Whether the user has unlimited generations.
  pub is_unlimited: bool,

  /// Whether the user is rate limited.
  pub is_rate_limited: bool,

  /// Whether the output will have a watermark.
  pub has_watermark: bool,
}

/// Error response for the image cost estimate endpoint.
#[derive(Serialize, Deserialize, ToSchema)]
pub struct EstimateImageCostError {
  pub success: bool,
  pub error_type: EstimateImageCostErrorType,
  pub error_message: String,
}

#[derive(Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum EstimateImageCostErrorType {
  InvalidProviderForModel,
  InvalidInput,
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn test_serialization_text_to_image() {
    let request = EstimateImageCostRequest {
      model: CommonImageModel::NanoBananaPro,
      provider: GenerationProvider::Artcraft,
      generation_mode: GenerationMode::TextToImage,
      aspect_ratio: None,
      resolution: None,
      quality: None,
      image_batch_count: None,
    };
    let serialized = serde_json::to_string(&request).unwrap();
    assert!(serialized.contains("\"generation_mode\":{\"type\":\"text_to_image\"}"));
  }

  #[test]
  fn test_serialization_image_edit() {
    let request = EstimateImageCostRequest {
      model: CommonImageModel::NanoBananaPro,
      provider: GenerationProvider::Artcraft,
      generation_mode: GenerationMode::ImageEdit { count: 2 },
      aspect_ratio: None,
      resolution: None,
      quality: None,
      image_batch_count: None,
    };
    let serialized = serde_json::to_string(&request).unwrap();
    assert!(serialized.contains("\"generation_mode\":{\"type\":\"image_edit\",\"count\":2}"));
  }
}

File: frontend/libs/tauri-api/src/lib/cost_estimate/EstimateImageCost.ts
import { invoke } from "@tauri-apps/api/core";
import { CommandResult, CommandSuccessStatus } from "../common/CommandStatus";
import {
  CommonImageModel,
  CommonAspectRatio,
  CommonQuality,
  CommonVideoResolution,
  ImageGenerationMode,
  GenerationProvider,
} from "@storyteller/api-enums";

export interface EstimateImageCostRequest {
  model: CommonImageModel;
  provider: GenerationProvider;
  generation_mode: ImageGenerationMode;
  aspect_ratio?: CommonAspectRatio;
  resolution?: CommonVideoResolution;
  quality?: CommonQuality;
  image_batch_count?: number;
}

export interface EstimateImageCostPayload {
  success: boolean;
  cost_in_credits?: number;
  cost_in_usd_cents?: number;
  is_free: boolean;
  is_unlimited: boolean;
  is_rate_limited: boolean;
  has_watermark: boolean;
}

export interface EstimateImageCostSuccess extends CommandResult {
  payload: EstimateImageCostPayload;
}

export interface EstimateImageCostErrorPayload {
  success: boolean;
  error_type: "invalid_provider_for_model" | "invalid_input";
  error_message: string;
}

export interface EstimateImageCostErrorResult extends CommandResult {
  error_details?: EstimateImageCostErrorPayload;
}

export type EstimateImageCostResult =
  | EstimateImageCostSuccess
  | EstimateImageCostErrorResult;

export const EstimateImageCost = async (
  request: EstimateImageCostRequest,
): Promise<EstimateImageCostResult> => {
  const result = await invoke("estimate_image_cost_command", { request });
  return result as EstimateImageCostResult;
};

export function isEstimateImageCostSuccess(
  r: EstimateImageCostResult,
): r is EstimateImageCostSuccess {
  return r.status === CommandSuccessStatus.Success;
}
